File: impls/clojure.2/src/mal/step1_read_print.clj
(ns mal.step1-read-print
  (:require
   [mal.printer]
   [mal.reader]))

(defn READ [s]
  (mal.reader/read-form s))

(defn EVAL [x]
  x)

(defn PRINT [s]
  (mal.printer/pr-str s))

(defn rep [input]
  (->
   input
   READ
   EVAL
   PRINT))

(loop []
  (print "user> ")
  (flush)
  (when-let [line (read-line)]
    (try
      (println (rep line))
      (catch Exception e
        (println (str "error:" (ex-message e)))
          ))
    (recur)))

File: impls/js2/step1_read_print.js

import { read_str } from "./reader.js";
import { pr_str } from "./printer.js";

const READ = (str) => read_str(str);

const EVAL = (ast) => ast;

const PRINT = (str) => pr_str(str);

const rep = (str) => PRINT(EVAL(READ(str)));

import * as readline from "node:readline";
import { stdin as input, nextTick, stdout as output } from "node:process";

const rl = readline.createInterface({ input, output });

const prompt = () => {
  rl.question(`user> `, (line) => {
    if (line === "") {
      console.log(`Bye!`);
      rl.close();
      return;
    }
    try {
      console.log(rep(line));
    } catch (e) {
      console.log(e.message);
    }
    nextTick(prompt);
  });
};

prompt();
